File: Shared/Blog.Core/Mapping/QueryableMappingExtensions.cs
// <copyright file="QueryableMappingExtensions.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

namespace Blog.Core.Mapping;

using AutoMapper.QueryableExtensions;
using System;
using System.Linq;
using System.Linq.Expressions;

/// <summary>
/// Queryable mapping extensions.
/// </summary>
public static class QueryableMappingExtensions
{
    /// <summary>
    /// To map.
    /// </summary>
    /// <typeparam name="TDestination">TDestination.</typeparam>
    /// <param name="source">source.</param>
    /// <param name="membersToExpand">membersToExpand.</param>
    /// <returns>IQueryable.</returns>
    public static IQueryable<TDestination> To<TDestination>(
        this IQueryable source,
        params Expression<Func<TDestination, object>>[] membersToExpand)
        => source == null
            ? throw new ArgumentNullException(nameof(source))
            : source.ProjectTo(null, membersToExpand);

    /// <summary>
    /// To map.
    /// </summary>
    /// <typeparam name="TDestination">TDestination.</typeparam>
    /// <param name="source">source.</param>
    /// <param name="parameters">parameters.</param>
    /// <returns>IQueryable.</returns>
    public static IQueryable<TDestination> To<TDestination>(
        this IQueryable source,
        object parameters)
        => source == null
            ? throw new ArgumentNullException(nameof(source))
            : source.ProjectTo<TDestination>((AutoMapper.IConfigurationProvider)parameters);
}

File: Shared/Blog.FSharp.Core/Mapping/QueryableMappingExtensions.fs
namespace Blog.FSharp.Core.Mapping

open System
open System.Linq
open System.Linq.Expressions
open AutoMapper
open AutoMapper.QueryableExtensions

[<AutoOpen>]
module QueryableMappingExtensions =

    type IQueryable with
        member this.To<'TDestination>([<ParamArray>] membersToExpand: Expression<Func<'TDestination, obj>>[]) =
            if isNull this then
                nullArg "source"
            this.ProjectTo(null, membersToExpand)

        member this.To<'TDestination>(parameters: obj) =
            if isNull this then
                nullArg "source"
            this.ProjectTo<'TDestination>(parameters :?> IConfigurationProvider)